src/xform_to_datalog/authorization_logic_datalog_facts.cc

```diff
@@ -0,0 +1,51 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#include "src/xform_to_datalog/authorization_logic_datalog_facts.h"
+
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <optional>
+#include <vector>
+
+#include "src/xform_to_datalog/authorization_logic.h"
+
+namespace raksha::xform_to_datalog {
+
+std::optional<AuthorizationLogicDatalogFacts>
+AuthorizationLogicDatalogFacts::create(
+  const std::string& program_dir, const std::string& program
+) {
+  auto output_dir = std::filesystem::temp_directory_path();
+  int res = generate_datalog_facts_from_authorization_logic(
+    program.c_str(), program_dir.c_str(), output_dir.c_str());
+  if (res) return std::nullopt;
+
+  // Read the file into a string.
+  std::filesystem::path output = output_dir / (program + ".dl");
+  std::ifstream file_stream(output);
+  if (!file_stream) return std::nullopt;
+
+  std::string datalog_program;
+  file_stream.seekg(0, std::ios::end);
+  datalog_program.resize(file_stream.tellg());
```

Comment:
Might be a bit more concise (and potentially save a `realloc` or similar under the hood) to just declare `datalog_program` a bit later and use the constructor that allocates a string with a provided size like so:

```C++
file_stream.seekg(0, std::ios::end);
std::string datalog_program(file_stream.tellg(), '\0');
```

Reply:
Also, we might want to check the value of `tellg` before sending it in here. According to the docs, `tellg` can return `-1` in at least some circumstances:

https://en.cppreference.com/w/cpp/io/basic_istream/tellg

But `resize` takes a `size_t`, which is unsigned, as its parameter. As such, there may exist a circumstance in which we may accidentally request a resize to `ULONG_MAX`.

Reply:
Good point about `tellg` and `gcount`. I made the changes. 

Won't `std::string datalog_program(file_stream.tellg(), '\0');` also generate instructions for zeroing out the bytes? It is not a big deal, but just wondering if it would essentially be the same.

Reply:
It would issue instructions to zero the array, yes. However, `resize` claims to *also* add characters to the array:  https://en.cppreference.com/w/cpp/string/basic_string/resize . It says that if resize expands the size of the string, it adds `CharT()` to the end. `cppreference` doesn't provide much guidance as to what `CharT()` evaluates to, but I'm guessing it's '\0'.

Reply:
Done.